src/main/scala/com/fulcrumgenomics/vcf/api/VcfSource.scala

```diff
@@ -0,0 +1,98 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.vcf.api
+
+import java.io.Closeable
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.commons.collection.SelfClosingIterator
+import htsjdk.samtools.util.CloseableIterator
+import htsjdk.variant.variantcontext.VariantContext
+import htsjdk.variant.vcf.VCFFileReader
+
+import scala.collection.IterableView
+
+/**
+  * Provides a reader over a source of VCF-like data that could be a VCF file or a BCF file. Has facilities
+  * for iterating over the entire stream of variants as well as querying by genomic location if an
+  * index is present.
+  *
+  * @param reader the underlying HTSJDK [[VCFFileReader]]
+  */
+class VcfSource private(private val reader: VCFFileReader) extends IterableView[Variant, VcfSource] with Closeable {
+  /** The header associated with the VCF being read. */
+  val header: VcfHeader = VcfConversions.toScalaHeader(reader.getFileHeader)
+
+  /**
+    * The type of iterator returned by both the [[iterator]] method as well as the [[query()]] method. Note that
+    * [[SelfClosingIterator]] both self-closes when it hits the end of the iterator, _and_ extends
+    * [[com.fulcrumgenomics.commons.collection.BetterBufferedIterator]].
+    */
+  type VariantIterator = SelfClosingIterator[Variant]
+
+  /** Wraps an iterator provided by HTSJDK into a SelfClosingIterator that transforms VariatnContexts into Variants. */
+  private def wrap(it: CloseableIterator[VariantContext]): VariantIterator = {
+    new SelfClosingIterator(
+      iter   = it.map(vc => VcfConversions.toScalaVariant(vc, header)),
+      closer = () => it.close())
+  }
+
+  /**
+    * Returns an iterator over the entire stream of variants. The returned iterator may be be closed by invoking
+    * `close()` on it, and will automatically close itself when exhausted.  Only a single iterator at a time
+    * is supported per [[VcfSource]], including iterators returned from [[query()]].
+    */
+  override def iterator: VariantIterator = wrap(reader.iterator())
+
+  /**
+    * Returns an iterator over variants overlapping the specified genomic region.
+    *
+    * The returned iterator may be be closed by invoking `close()` on it, and will automatically close itself
+    * when exhausted.  Only a single iterator at a time is supported per [[VcfSource]], including iterators
+    * returned from [[iterator()]].
+    */
+  def query(chrom: String, start: Int, end: Int): Iterator[Variant] = wrap(reader.query(chrom, start, end))
```

Comment:
Do we also want a query method that takes in > 1 range and provides a single iterator of unique variants coming back?  This might be tricky to do as the functionality isn't in the lower level library _and_ I think we can only have one active query at a time.

Reply:
Nah, it shouldn't be hard to do it manually for now.